reframe/frontend/statistics.py

```diff
@@ -1,89 +1,98 @@
+from itertools import chain
+
 import reframe.core.debug as debug
+import reframe.core.exceptions as exc
 
 
 class TestStats:
     """Stores test case statistics."""
 
-    def __init__(self, tasks=[]):
-        if not hasattr(tasks, '__iter__'):
-            raise TypeError('expected an iterable')
-
-        # Store test cases per partition internally
-        self._tasks_bypart = {}
-        for t in tasks:
-            partition = t.check.current_partition
-            partname = partition.fullname if partition else 'None'
-            tclist = self._tasks_bypart.setdefault(partname, [])
-            tclist.append(t)
+    def __init__(self):
+        self._tasks = []
+        self._try_num = -1
 
     def __repr__(self):
         return debug.repr(self)
 
-    def num_failures(self, partition=None):
-        num_fails = 0
-        if partition:
-            num_fails += len([
-                t for t in self._tasks_bypart[partition] if t.failed
-            ])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_fails += len([t for t in tclist if t.failed])
-
-        return num_fails
-
-    def num_failures_stage(self, stage):
-        num_fails = 0
-        for tclist in self._tasks_bypart.values():
-            num_fails += len([t for t in tclist if t.failed_stage == stage])
-
-        return num_fails
-
-    def num_cases(self, partition=None):
-        num_cases = 0
-        if partition:
-            num_cases += len(self._tasks_bypart[partition])
-        else:
-            # count all failures
-            for tclist in self._tasks_bypart.values():
-                num_cases += len(tclist)
-
-        return num_cases
-
-    def failure_report(self):
+    def next_try(self, try_num):
+        # TODO: RAISE DIFFERENT error?
```

Comment:
I disagree with this interface. You're giving too much power to the callers, without even needing it. Then you have to deal with bad inputs. This should be very very simple:

```python
def addrun(self):
    self._tasks.append([])
    self._current_run += 1
```

Reply:
No, my intention is not to give power to the caller, I just try to ensure that the counter the `statistics` realizes if his counter is out of sync with the runners counter (as the runner has the real counter of the retries and in statistics we have only a duplication of it).

Reply:
No, keep the interface simple. The `Runner` should make sure to call correctly this method and whenever necessary.

Reply:
Done.